pennylane_lightning/core/src/measurements/MeasurementsBase.hpp

```diff
@@ -479,12 +480,14 @@ template <class StateVectorT, class Derived> class MeasurementsBase {
             StateVectorT sv(data_storage.data(), data_storage.size());
             obs.applyInPlaceShots(sv, eigenValues, obs_wires);
             Derived measure(sv);
-            samples = measure.generate_samples(num_shots);
+            samples = measure.generate_samples(num_shots,
+                                               static_cast<Derived *>(this));
         } else {
             StateVectorT sv(_statevector);
             obs.applyInPlaceShots(sv, eigenValues, obs_wires);
             Derived measure(sv);
-            samples = measure.generate_samples(num_shots);
+            samples = measure.generate_samples(num_shots,
```

Comment:
What about calling `set_PRNG` on `Derived` here rather than passing down `this`?

Reply:
I'm not sure about your question, someone from perf would know how to answer it. 

> Calling set_PRNG is assuming all derived classes have implemented that method.

What if the base class implemented those methods?

Reply:
That would be the simplest solution if there is no problem with being explicit about the usage of PRNGs for generating samples at the base class. @maliasadi what do you think?

Reply:
Checking the measurement base class, I realized there exists a PRNG for the internal use of the measure. Adding a second PRNG would be confusing. So, I have simplified the design and added a device seed to the base measurement class. This member variable serves as the seed of the base class when provided. If not provided, a random seed is used. Therefore, there is no need anymore to have a copy of the PRNG in the derived measure class. With the new design, all methods calling the sample generation routine will take into account the base class seed. The simulator is the one setting the measure seed using its own PRNG right after creating it the measure. The sample generation routine in the derived measure class is the one activating the seed.

Reply:
And we are no longer passing the whole derived measurement class as an argument :)